### PR TITLE
feat(models): Mistral3.1 VLM pytorch backend support

### PR DESCRIPTION
**Pull request by 2ez4bz**
# [TRTLLM-5341]feat(models): Mistral3.1 VLM pytorch backend support

## Description

This commit:
- defines a new `Mistral3VLM` class that replicates the functionality
 `Mistral3ForConditionalGeneration` in the transformers library. This
  allows using models such as `Mistral-Small-3.1-24B-Instruct-2503`
  in TRTLLM's pytoch backend.
- adds a new `modeling_pixtral.py` file replicating the functionality
  from `transformers` but using TRTLLM modules where appropriate.
- extends an existing E2E test for the new model with image inputs.

## Test Coverage

New test case added to `tests/integration/defs/test_e2e.py::test_ptp_quickstart_multimodal`.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_mistral.py [hunk 1]
```diff
@@ -187,3 +195,188 @@ def __init__(
...
+        self._vision_tower = ref_model.vision_tower.to(self._device)
+        self._vision_feature_layer = model_config.vision_feature_layer
+        self._multi_modal_projector = ref_model.multi_modal_projector.to(
+            self._device)
+
+        self._processor = AutoProcessor.from_pretrained(model_path,
+                                                        use_fast=False)
+
+    @torch.inference_mode()
+    def __call__(
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_mistral.py:237]: I've been told by @brb-nv that it would be preferrable to move this into the `Mistral3VLM.forward` method. Can reviewers confirm whether that's true?

File: tensorrt_llm/inputs/utils.py [hunk 2]
```diff
@@ -217,6 +219,9 @@ class MultimodalPlaceholderPlacement(enum.Enum):
     "mllama": MultimodalPlaceholderPlacement.BEFORE_TEXT,
     "hyperclovax_vlm": MultimodalPlaceholderPlacement.AFTER_TEXT,
     "gemma3": MultimodalPlaceholderPlacement.BEFORE_TEXT,
+    # TODO: double check if this is correct. I'm assuming this based on the prompt from vLLM
```
Comment (2ez4bz) [inline tensorrt_llm/inputs/utils.py:222]: @dongxuy04 @yuxianq Calling this out - please let me know how this is determined 🙏

File: tensorrt_llm/inputs/utils.py [hunk 3]
```diff
@@ -247,6 +252,11 @@ def retrieve_multimodal_placeholder(model_type: str, modality: str,
                     '<|im_start|>user (vector)\n<|dummy3|><|im_end|>\n' + \
                     '<|im_start|>image/aux\n다음 중 ocr은 사진에서 검출된 글자이고, lens_keyword는 사진에서 추출된 keyword와 bbox 위치입니다.' + \
                     'bbox는 0~1 사이로 정규화된 [x1, y1, x2, y2]의 형태입니다. 참고하여 답변하세요. {"ocr": "", "lens_keywords": "", "lens_local_keywords": ""}'
+        elif model_type in SUPPORTED_MISTRAL_IMAGE_MODEL_GROUP:
+            # ? How to make sure this is correct? I am purely basing this off of:
```
Comment (2ez4bz) [inline tensorrt_llm/inputs/utils.py:256]: @dongxuy04 @yuxianq Calling this out 🙏

File: tests/unittest/_torch/modeling/test_modeling_mistral.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+)
+@torch.no_grad()
+def test_mistral3_vlm_allclose_to_hf(mistral_small_3_1_24b_config, backend, use_cuda_graph):
+    metadata_cls = attention_utils.get_attention_backend(backend).Metadata
+
+    torch.random.manual_seed(0)
+    config_dict = mistral_small_3_1_24b_config
+    # 24B * sizeof(float16) plus some extra for activations
+    # times 2, since we'll need 2 of these
+    mem_for_full_model = int(2.1 * 24 * 2 ** (30) * 2)
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_modeling_mistral.py:243]: The original code I swiped this from in `test_modeling_llama.py` had a `* 4` at the end, even though the code comment above also said `times 2`. @yechank-nvidia can you confirm which one is correct?

File: tensorrt_llm/inputs/utils.py (hunk 2 above)
Comment (yuxianq) [reply tensorrt_llm/inputs/utils.py:222]: @rakib-hasan Can you help to review this change?

File: tensorrt_llm/inputs/utils.py (hunk 3 above)
Comment (yuxianq) [reply tensorrt_llm/inputs/utils.py:256]: @rakib-hasan Can you help to review this change?

File: tensorrt_llm/_torch/models/modeling_mistral.py (hunk 1 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_mistral.py:237]: So in the InputProcessor everything except the Vision Encoder's forward() should be placed.

**Review by yechank-nvidia — COMMENTED** [review]
- InputProcessor: Tokenize & AutoProcessor's process() => send pixel values & mm_extra_data(e.g. image metadata)
- Vision Encoder & LLM: Vision Encoder forward() & LLM forward()

You need to have separate class of Vision Encoder and if possible, we need to use TRT-LLM's format of Vision Encoder instead of using HF's one. The reason is that we will have duplicate Vision Encoders on tp > 1 scenario on HF. 

You can refer this [PR](https://github.com/NVIDIA/TensorRT-LLM/pull/5522/files). Take a look at how modeling_hyperclovax.py and modeling_qwen2vl.py is refactored.

Also, seems like you also need to send mm_extra_data(e.g. image_size) to the Vision Encoder, then you would need to wait for upper PR to be merged.

File: tensorrt_llm/inputs/utils.py (hunk 3 above)
Comment (2ez4bz) [reply tensorrt_llm/inputs/utils.py:256]: I think this is correct based on this? https://github.com/mistralai/mistral-common/blob/26a6bb3a07ee0b78a3808f2797f23e1d28514b93/src/mistral_common/tokens/tokenizers/base.py#L60

File: tests/unittest/_torch/modeling/test_modeling_pixtral.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
+import contextlib
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_modeling_pixtral.py:1]: This entire file needs more work.

File: tensorrt_llm/_torch/models/modeling_mistral.py [hunk 6]
```diff
@@ -187,3 +195,188 @@ def __init__(
...
+                image_sizes=processed["image_sizes"].to(self._device),
+            )
+            extra_processed_inputs = {"mm_embedding": image_features}
+
+        return input_ids, extra_processed_inputs
+
+    # Original implementation:
+    # https://github.com/huggingface/transformers/blob/v4.51.3/src/transformers/models/mistral3/
+    # modeling_mistral3.py#L341
+    def _get_image_features(
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_mistral.py]: Done.

File: tensorrt_llm/_torch/models/modeling_pixtral.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+    @torch.inference_mode()
+    def forward(
+        self,
+        pixel_values: torch.Tensor,
+        image_sizes: torch.Tensor,
+        output_hidden_states: Optional[bool] = None,
+    ):
+        with torch.autocast(device_type="cuda", dtype=self.config.torch_dtype):
+            patch_embeds = self.patch_conv(pixel_values)
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_pixtral.py:193]: Yes. Layers that can not be defined under TRT-LLM's logic, we should duplicate weights on `tp>1`. (For example, [LlavaNextMultiModalProjector](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/models/modeling_llava_next.py#L62))

File: tensorrt_llm/_torch/models/modeling_mistral.py [hunk 8]
```diff
@@ -187,3 +197,306 @@ def __init__(
...
+        )
+
+        image_features = self._multi_modal_projector(image_outputs.squeeze(0),
+                                                     image_sizes)
+        return image_features
+
+
+# Original implementation:
+# https://github.com/huggingface/transformers/blob/v4.51.3/src/transformers/models/mistral3/modeling_mistral3.py#L66
+class Mistral3PatchMerger(torch.nn.Module):
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_mistral.py:394]: If you don't modify any code, can you import it by `from transformers.models.mistral3.modeling_mistral3 import Mistral3PatchMerger`?

File: tests/integration/defs/test_e2e.py [hunk 9]
```diff
@@ -1920,9 +1920,15 @@ def test_ptp_quickstart_advanced_mixed_precision(llm_root, llm_venv):
     ("llava-v1.6-mistral-7b", "llava-v1.6-mistral-7b-hf"),
     ("qwen2-vl-7b-instruct", "Qwen2-VL-7B-Instruct"),
     ("qwen2.5-vl-7b-instruct", "Qwen2.5-VL-7B-Instruct"),
+    ("mistral-small-3.1-24b-instruct", "Mistral-Small-3.1-24B-Instruct-2503"),
 ])
 def test_ptp_quickstart_multimodal(llm_root, llm_venv, model_name, model_path,
                                    modality):
+    if modality == "video" and "mistral" in model_name:
```
Comment (yechank-nvidia) [inline tests/integration/defs/test_e2e.py:1927]: You don't need to do make a branch here. Since, we are doing test by giving them the supported modality like `test_ptp_quickstart_multimodal[NVILA-8B-FP16-vila/NVILA-8B-video]`

You can make the test running [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/integration/test_lists/qa/examples_test_list.txt#L523), and [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/integration/test_lists/test-db/l0_l40s.yml#L27).

File: tensorrt_llm/_torch/models/modeling_pixtral.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+            )
+
+            hidden_states = layer_outputs
+
+        return hidden_states
+
+
+# Original implementation:
+# https://github.com/huggingface/transformers/blob/v4.51.3/src/transformers/models/pixtral/modeling_pixtral.py#L440
+# @reviewer: Should I be calling `register_auto_model` for this?
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_pixtral.py:153]: Not sure people will use PixtralVisionModel on other models, but I believe registering is no harm. [CLIPVisionModel](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/models/modeling_clip.py#L177) also did this.

File: tensorrt_llm/_torch/models/modeling_mistral.py [hunk 11]
```diff
@@ -187,3 +205,322 @@ def __init__(
...
+    @torch.inference_mode()
+    def __call__(
+        self, inputs: TextPrompt, sampling_params: SamplingParams
+    ) -> Tuple[List[int], Optional[ExtraProcessedInputs]]:
+        images = inputs.get("multi_modal_data", {}).get("image")
+        if images is not None:
+            # Although the `PixtralProcessor` supports PIL images, numpy arrays, and torch tensors,
+            # they are expected to be in the [0, 255] range, not [0, 1]. This is a heuristic to
+            # convert them to the expected range.
+            if isinstance(images[0], torch.Tensor) and images[0].mean() < 1.0:
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_mistral.py:241]: The `do_rescale` option is set to `True` by default when loading the config from Mistral's checkpoint:

```
>>> processor = transformers.PixtralProcessor.from_pretrained(os.path.join(LLM_MODELS_ROOT), "Mistral-Small-3.1-24B-Instruct-2503")
>>> print(processor.image_processor)
PixtralImageProcessorFast {
  "crop_size": null,
  "data_format": "channels_first",
  "default_to_square": true,
  "device": null,
  "do_center_crop": null,
  "do_convert_rgb": true,
  "do_normalize": true,
  "do_rescale": true,
  "do_resize": true,
  "image_mean": [
    0.48145466,
    0.4578275,
    0.40821073
  ],
  "image_processor_type": "PixtralImageProcessorFast",
  "image_std": [
    0.26862954,
    0.26130258,
    0.27577711
  ],
  "input_data_format": null,
  "patch_size": 14,
  "processor_class": "PixtralProcessor",
  "resample": 3,
  "rescale_factor": 0.00392156862745098,
  "return_tensors": null,
  "size": {
    "longest_edge": 1540
  }
}
```

Per pixtral's docstring [here](https://github.com/huggingface/transformers/blob/v4.51.3/src/transformers/models/pixtral/image_processing_pixtral.py#L232), the resizing is done by looking at either the longest or shortest edge, and resizing while trying to preserve the aspect ratio + respecting multiples of the `patch_size`.

Extra padding is applied when there are multiple images passed to the processor itself. But, individual images are not resized to a "universal" `(height, width)` when passed individually:

image

File: tests/integration/defs/test_e2e.py [hunk 12]
```diff
@@ -2031,6 +2034,18 @@ def test_ptp_quickstart_multimodal(llm_root, llm_venv, model_name, model_path,
...
+        "mistral-small-3.1-24b-instruct": {
+            "image": [
+                [
+                    "dramatic", "seascape", "stormy", "turbulent", "waves",
+                    "rough"
+                ],
+                ["scenic", "rock", "landscape", "snow", "formation"],
+                ["highway", "traffic", "directions", "lanes", "Jurong"],
+            ],
+            "video":
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tests/integration/defs/test_e2e.py:2046]: Removed.

File: tensorrt_llm/_torch/models/modeling_mistral.py [hunk 13]
```diff
@@ -187,3 +205,322 @@ def __init__(
...
+            #    `PixtralProcessor`.
+            # 2. After each `pixel_values` tensor has gone through the vision tower's `patch_conv` layer,
+            #    they are divided into patches that are then concatenated in order to treat them as a
+            #    single "sequence" in the vision tower's attention layers, so some form of batching still
+            #    happens in the vision tower.
+            image_features = [
+                self._get_image_features(**x.multimodal_data["image"])
+                for x in multimodal_params
+            ]
+            input_ids, inputs_embeds = fuse_input_embeds(
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_mistral.py:363]: Done.

File: tensorrt_llm/_torch/models/modeling_mistral.py [hunk 14]
```diff
@@ -187,3 +205,322 @@ def __init__(
...
+        attn_metadata: AttentionMetadata,
+        input_ids: Optional[torch.LongTensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
+        return_context_logits: bool = False,
+        **kwargs,
+    ) -> torch.Tensor:
+        """Forward method."""
+        multimodal_params = kwargs.get("multimodal_params", [])
+        if len(multimodal_params) > 0:
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_mistral.py:348]: Added.

File: tensorrt_llm/inputs/utils.py (hunk 2 above)
Comment (2ez4bz) [reply tensorrt_llm/inputs/utils.py]: Thanks for the pointer. Added a comment here to point to it, plus one in the docstring for `tensorrt_llm/_torch/models/modeling_mistral.py::Mistral3VLM`.

File: tensorrt_llm/inputs/utils.py (hunk 3 above)
Comment (2ez4bz) [reply tensorrt_llm/inputs/utils.py]: That is correct; however, we have to return some placeholder, as otherwise the pipeline does not work (I tried returning `None` here, and force the chat template to be applied - it does not "pad" the input prompt with `[IMG]` tokens).

File: tests/unittest/_torch/modeling/test_modeling_mistral.py (hunk 4 above)
Comment (yechank-nvidia) [reply tests/unittest/_torch/modeling/test_modeling_mistral.py:259]: Talked offline.

**Review by yechank-nvidia — APPROVED** [review]
LGTM. Thx for the work!

**Review by lucaslie — APPROVED** [review]
Blind-approving on behalf of @NVIDIA/trt-llm-torch-devs since PR has already been approved and has appropriate test coverage
